Upgrade Java dependencies and fix instrumentation test

mrsarm opened the pull request:
Upgrade building dependencies (https://github.com/medic/cht-android/issues/174), and fix broken CI because gamma dev environment started to use French as default language (https://github.com/medic/cht-android/issues/191).

- Upgrade dependencies (https://github.com/medic/cht-android/issues/174):
  - Min Java ~6~ → 8
  - Gradle ~6.5~ → 7.1
  - Android Gradle Plugin ~4.1.3~ → 4.2.2
  - Other minor upgrades of building and testing libraries
-  Fix login test broken when development environment use French as default language instead of English (https://github.com/medic/cht-android/issues/191).
- Fix lint configuration that now works again when used with the XWalk builds.
- Refactor some code replacing old anonymous Java classes with lambda style (Java 8
- Improvements in the _Makefile_ scripts, add ability to set the flavor in some of the targets. 
- Fix indentation in `build.gradle`.
- Add more info in the _"Development"_ section in the README.

latin-panda commented on README.md:
```diff
@@ -12,12 +12,13 @@ To help you pick which APK to install you can find information about the version
+| Rendering engine | Instruction set | Android version | Notes                                                       |
+|------------------|-----------------|-----------------|-------------------------------------------------------------|
+| `webview`        | `arm64-v8a`     | 10+             | Preferred. Use this APK if possible.                        |
 | `webview`        | `armeabi-v7a`   | 10+             | Built but not compatible with any devices. Ignore this APK. |
```
Unrelated question to this PR:

Why do we build the `armeabi-v7a` APK if it can't be used? 🤔 Can we ask to remove it from the build to avoid confusion?

garethbowen replied:
I think it automatically builds every combination of every dimension, so I don't think we can stop it. We _could_ delete it when the build finishes with a custom script. It's probably not worth spending time on at this point as it will almost certainly go away when we solve [this issue](https://github.com/medic/cht-android/issues/164) because after that we expect to build one bundle instead of a series of apks.

latin-panda commented on README.md:
```diff
@@ -91,22 +92,39 @@ This release changes the way in which location data is collected to better align
-2. Clone the repo
-3. Plug in your phone. Check it's detected with `adb devices`
-4. Execute: `make` (will also push app unto phone)
+1. Install the [Android SDK](https://developer.android.com/studio#command-tools) and Java 8+ (it works with the OpenJDK versions).
```
Do we need to specify that we're using gradle 7?

mrsarm replied:
No because the `gradlew` script takes care of that, downloading Gradle 7.1 in the user space the first time is executed, and the build is still compatible with older versions of Gradle, but I'll add a note about it.

latin-panda commented on src/androidTestUnbrandedWebviewDebug/java/org/medicmobile/webapp/mobile/SettingsDialogActivityTest.java:
```diff
@@ -121,7 +123,7 @@ public void testLoginScreen() throws Exception {
 		Thread.sleep(4000);//TODO: use better ways to handle delays - takes longer with emulators
 		onWebView().withElement(findElement(Locator.CSS_SELECTOR, "p.error.incorrect"))
-				.check(webMatches(getText(), containsString(ERROR_INCORRECT)));
+				.check(webMatches(getText(), matchesPattern(ERROR_INCORRECT)));
```
Is it possible to select the language in the login page instead? 
Because if we decide to add more gamma testing, it's going to be difficult to maintain 2 labels every time.

latin-panda commented on src/main/java/org/medicmobile/webapp/mobile/AlertDialogUtils.java:
```diff
@@ -31,11 +30,7 @@ public static void show(
-					dialog.dismiss();
-				}
-			})
+			.setNeutralButton(okLabel, (dialog, which) -> dialog.dismiss())
```
✨ So nice to do this! ✨

latin-panda commented on src/xwalk/java/org/medicmobile/webapp/mobile/EmbeddedBrowserActivity.java:
```diff
@@ -253,19 +253,17 @@ SmsSender getSmsSender() {
+			// to run JS, in which case we should switch to the second
+			// block.
+			// On switching to XWalkView, we assume the same applies.
+			if(true) { // NOPMD
```
I know this code was there before.... but if it's always true, can we remove the `else` section ?
Do you think this big comment can be placed before the `container.post`? then we can have 1 line function 🤔

mrsarm replied:
Yes didn't notice the if inside when refactoring.

latin-panda reviewed (changes requested):
Nice Job ✨ I'm so happy to have this upgrade, it'll provide a lot of flexibility!!
I've left some comments and questions in line, nothing big.
Also, I've done a quick smoke test in Android 11 and 9 and look okay, but I'd recommend to test in older devices, I think Bede has an Android 6.

mrsarm commented on README.md:
```diff
@@ -91,22 +92,39 @@ This release changes the way in which location data is collected to better align
-2. Clone the repo
-3. Plug in your phone. Check it's detected with `adb devices`
-4. Execute: `./gradlew connected[Flavor]WebviewDebugAndroidTest` . Eg `./gradlew connectedUnbrandedWebviewDebugAndroidTest` or `./gradlew connectedMedicmobilegammaWebviewDebugAndroidTest`. At the moment we have tests only in these 2 flavors: unbranded and medicmobilegamma. To avoid failures running the tests, previous versions of the app should be uninstalled first, otherwise an `InstallException ...
+These tests run on your device.
```
Yes we only support physical devices from Gradle and Make to run and test the app.

mrsarm commented:
@latin-panda please review again, here are the changes since your last review:

- Refactor and improvements of the Makefile, to support the flavor argument in some of the targets, and move boilerplate configuration to env variable.
- Moved some code that was only used by the Webview version that caused the linter to fail when executing the check for the Xwalk version (`make flavour=UnbrandedXWalk lint`).
- Changed the documentation using `make` command instead of the Gradle versions


Also I forget to mention that one of the goal of this PR was to remove the **JCenter** repository, but we cannot do that due that we have two dependencies that are only hosted in JCenter, but according with [this](https://jfrog.com/blog/into-the-sunset-bintray-jcenter-gocenter-and-chartcenter/), JFrog, the repo's owner, quote:

> _...will keep JCenter as a read-only repository indefinitely. Our customers and the community can continue to rely on JCenter as a reliable mirror for Java packages._

latin-panda reviewed (commented):
Nice job!! ✨ I think the makefile commands are super easy now, thanks!!
I left a couple of suggestion in the documentation, the rest is fine!

latin-panda commented on README.md:
```diff
@@ -91,22 +92,54 @@ This release changes the way in which location data is collected to better align
+
+## Flavor selection
+
+Some *Make* targets support the flavor and rendering engine as `make flavour=[Flavor][Engine] TASK`, where `[Flavor]` is the branded version with the first latter capitalized and the engine is either `Webview` or `Xwalk`. The default value for `flavor` is `UnbrandedWebview`, e.g. executing `make deploy` will assemble and install that flavor, while executing `make flavour=MedicmobilegammaXwalk dep ...
```
I'm wondering if the `make` argument `flavour=` should be `flavor`... looking at the android docs, it appears that they use `flavor` 🤔

mrsarm replied:
Yes I guess is an old typo, fixing...

mrsarm commented on README.md:
```diff
@@ -91,22 +92,39 @@ This release changes the way in which location data is collected to better align
+
+To only execute the **linter checks**, execute: `./gradlew androidCheckstyle`.
+
+The [Android Studio](https://developer.android.com/studio) can be used to build and launch the app instead. Be sure to select the right flavor and rendering engine from the _Build Variants_ dialog (see [Build and run your app](https://developer.android.com/studio/run)). To launch the app in an emulator, you need to uncomment the code that has the strings for the `x86` or the `x86_64` architecture ...
```
I made it a sub section, but the links provided should be enough for the user to go deeper of how to use Android Studio and how to setup emulators. I only wanted to mention about the selection of the variant and the commented emulators' arch because are the typical question that I had and others had in the past when starting to work with the app.

mrsarm commented on src/androidTestUnbrandedWebviewDebug/java/org/medicmobile/webapp/mobile/SettingsDialogActivityTest.java:
```diff
@@ -92,7 +92,7 @@ public void testLoginScreen() throws Exception {
 				.atPosition(2);
 		linearLayout.perform(click());
-		Thread.sleep(10000);//TODO: use better ways to handle delays
+		Thread.sleep(12000);	//TODO: use better ways to handle delays
```
`Thread.sleep(..)` is a really bad option when running multiple threads, but here we only run once at a time, so for now I'll keep it like that.

latin-panda reviewed (approved):
LGTM! This PR includes great improvements thanks!! ☀️🎉

mrsarm commented:
Going to merge without moving it to AT because only build and testing dependencies were updated (anyway I did many test locally with both XWalk and Webview flavors).
